modules/ximgproc/src/paillou_filter.cpp

```diff
@@ -0,0 +1,449 @@
+         return *this;
+    };
+};
+
+
+class ParallelGradientPaillouYRows: public ParallelLoopBody
+{
+private:
+    Mat &img;
+    Mat &dst;
+    double a;
+    double w;
+    bool verbose;
+
+public:
+    ParallelGradientPaillouYRows(Mat& imgSrc, Mat &d,double aa,double ww):
+        img(imgSrc),
+        dst(d),
+        a(aa),
+        w(ww),
+        verbose(false)
+    {}
+    void Verbose(bool b){verbose=b;}
+    virtual void operator()(const Range& range) const
+    {
+        CV_Assert(img.depth()==CV_32FC1);
+        if (verbose)
+            std::cout << getThreadNum()<<"# :Start from row " << range.start << " to "  << range.end-1<<" ("<<range.end-range.start<<" loops)" << std::endl;
+        float *iy,*iy0;
+        int tailleSequence=(img.rows>img.cols)?img.rows:img.cols;
+        Mat matIym(1,tailleSequence,CV_64FC1),  matIyp(1,tailleSequence,CV_64FC1);
+        double *iym=(double*)matIym.ptr(0), *iyp=(double*)matIyp.ptr(0);
+        int cols=img.cols;
+
+        // Equation 13 p193
+        double                d=(1-2*exp(-a)*cosh(w)+exp(-2*a))/(2*a*exp(-a)*sinh(w)+w*(1-exp(-2*a)));
+        double                c1=a*d;
+        double                c2=w*d;
+        // Equation 12 p193
+        double                b1=-2*exp(-a)*cosh(w);
+        double                b2=exp(-2*a);
+        // Equation 14 p193
+        double                a0p=c2;
+        double                a1p=(c1*sinh(w)-c2*cosh(w))*exp(-a);
+        double                a1m=a1p-c2*b1;
+        double                a2m=-c2*b2;
+
+        for (int i=range.start;i<range.end;i++)
+            {
+            iy0 = ((float*)img.ptr(i));
+            int j=0;
+            iyp[0] = a0p*iy0[0] ;
+            iyp[1] = a0p*iy0[1] + a1p*iy0[0] - b1*iyp[0];
+            iy0 += 2;
+            for (j=2;j<cols;j++,iy0++)
+                iyp[j] = a0p*iy0[0] + a1p*iy0[-1] - b1*iyp[j-1] - b2*iyp[j-2];
+            iy0 = ((float*)img.ptr(i))+cols-1;
+            iym[cols-1] = 0;
+            iy0--;
+            iym[cols-2] = a1m*iy0[1]  - b1*iym[cols-1];
+            iy0--;
+            for (j=cols-3;j>=0;j--,iy0--)
+                iym[j] = a1m*iy0[1] + a2m*iy0[2] - b1*iym[j+1] - b2*iym[j+2];
+            iy = ((float*)dst.ptr(i));
+            for (j=0;j<cols;j++,iy++)
+                *iy = (float)(iym[j]+iyp[j]);
+            }
+
+    };
+    ParallelGradientPaillouYRows& operator=(const ParallelGradientPaillouYRows &) {
+         return *this;
+    };
+};
+
+
+class ParallelGradientPaillouXCols: public ParallelLoopBody
+{
+private:
+    Mat &img;
+    Mat &dst;
+    double a;
+    double w;
+    bool verbose;
+
+public:
+    ParallelGradientPaillouXCols(Mat& imgSrc, Mat &d,double aa,double ww):
+        img(imgSrc),
+        dst(d),
+        a(aa),
+        w(ww),
+        verbose(false)
+    {}
+    void Verbose(bool b){verbose=b;}
+    virtual void operator()(const Range& range) const
+    {
+        CV_Assert(img.depth()==CV_32FC1);
```

Comment:
You mean   :
*1 ParallelGradientPaillouXRows and  ParallelGradientPaillouXCols support different varieties of Mat types
*2 ParallelGradientPaillouYCols and ParallelGradientPaillouYRows support different varieties of Mat types

Yes because there is two steps in thoses filter. first step ParallelGradientPaillouXRows  (or ParallelGradientPaillouYCols ) process orignal image and give a Mat<float>. Second step process previous Mat<float> and give Mat<float> as result.